### PR TITLE
feat(native/FileUpload): add Native FileUpload component

### PR DESCRIPTION
**Pull request by shridhar52war**
- Props for File Upload Component

```
{
  progress: number, // to indicate the progress of the upload
  title: string, 
  file: string, // file name
  accept: array, // Accepts: ["allFiles", "audio", "csv", "doc", "docx", "images", "pdf", "plainText", "ppt", "pptx", "video", "xls", "xlsx", "zip"]
  errorText: string,
  helpText: string,
  onError: func, // Callback is called when the native module is not able to pick the file
  onFileUpload: func, // Callback is called when file is successfully picked by native module,
  onRemove: func, //callback when user clicks on cross icon
}
```
- Currently Progress Bar component resides within FileUpload Component, will move it to separate component once the APIs are finalized.
- ProgressBar Pops 
```
  {
    color: "", // color of the progress bar[blade colors]
    error: string, 
    size: oneOf(['small', 'medium', 'large']),
    progress: PropTypes.number

  }
```
- NOTE: Have not written test cases for it, wanted to confirm on the API. 🙈

### CODE REVIEW

File: src/molecules/FileUpload/FileUpload.native.js [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+FileUpload.propTypes = {
+  progress: PropTypes.number,
+  title: PropTypes.string,
+  file: PropTypes.string,
+  accept: PropTypes.arrayOf(Object.keys(DocumentPicker.types)), // Accepts: ["allFiles", "audio", "csv", "doc", "docx", "images", "pdf", "plainText", "ppt", "pptx", "video", "xls", "xlsx", "zip"]
+  errorText: PropTypes.string,
+  helpText: PropTypes.string,
+  onError: PropTypes.func,
+  onFileUpload: PropTypes.func,
```
Comment (chaitanyadeorukhkar) [inline src/molecules/FileUpload/FileUpload.native.js:149]: Upload is misleading here
```suggestion
  onFileSelected: PropTypes.func,
```

File: src/molecules/FileUpload/FileUpload.native.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+      ) : null}
+    </View>
+  );
+};
+
+FileUpload.propTypes = {
+  progress: PropTypes.number,
+  title: PropTypes.string,
+  file: PropTypes.string,
+  accept: PropTypes.arrayOf(Object.keys(DocumentPicker.types)), // Accepts: ["allFiles", "audio", "csv", "doc", "docx", "images", "pdf", "plainText", "ppt", "pptx", "video", "xls", "xlsx", "zip"]
```
Comment (chaitanyadeorukhkar) [inline src/molecules/FileUpload/FileUpload.native.js:145]: Seems clearer
```suggestion
  fileTypes: PropTypes.arrayOf(Object.keys(DocumentPicker.types)), // Accepts: ["allFiles", "audio", "csv", "doc", "docx", "images", "pdf", "plainText", "ppt", "pptx", "video", "xls", "xlsx", "zip"]
```

File: src/molecules/FileUpload/FileUpload.native.js [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        </Space>
+      ) : null}
+    </View>
+  );
+};
+
+FileUpload.propTypes = {
+  progress: PropTypes.number,
+  title: PropTypes.string,
+  file: PropTypes.string,
```
Comment (chaitanyadeorukhkar) [inline src/molecules/FileUpload/FileUpload.native.js:144]: What exactly is this?

File: src/molecules/FileUpload/FileUpload.native.js [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  accept: PropTypes.arrayOf(Object.keys(DocumentPicker.types)), // Accepts: ["allFiles", "audio", "csv", "doc", "docx", "images", "pdf", "plainText", "ppt", "pptx", "video", "xls", "xlsx", "zip"]
+  errorText: PropTypes.string,
+  helpText: PropTypes.string,
+  onError: PropTypes.func,
+  onFileUpload: PropTypes.func,
+  onRemove: PropTypes.func,
```
Comment (chaitanyadeorukhkar) [inline src/molecules/FileUpload/FileUpload.native.js:150]: Something consistent with onFileSelected?
```suggestion
  onFileDismissed: PropTypes.func,
```

**Review by chaitanyadeorukhkar — CHANGES_REQUESTED** [review]

File: src/molecules/FileUpload/FileUpload.native.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+};
+
+FileUpload.propTypes = {
+  progress: PropTypes.number,
+  title: PropTypes.string,
+  file: PropTypes.string,
+  accept: PropTypes.arrayOf(Object.keys(DocumentPicker.types)), // Accepts: ["allFiles", "audio", "csv", "doc", "docx", "images", "pdf", "plainText", "ppt", "pptx", "video", "xls", "xlsx", "zip"]
+  errorText: PropTypes.string,
+  helpText: PropTypes.string,
+  onError: PropTypes.func,
```
Comment (shubhnik) [inline src/molecules/FileUpload/FileUpload.native.js:148]: error when? maybe something like `onFileSelectionError`?

File: src/molecules/FileUpload/FileUpload.native.js [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+import Icon from '../../atoms/Icon';
+import ProgressBar from './ProgressBar';
+
+const MAX_PROGRESS_VALUE = 100;
+
+const DashedButton = styled(TouchableOpacity)`
+  background-color: ${(props) => props.theme.colors.background[200]};
+  border: 1px dashed ${(props) => props.theme.colors.tone[900]};
+  border-radius: 2px;
+  padding: 10px 12px;
```
Comment (shubhnik) [inline src/molecules/FileUpload/FileUpload.native.js:20]: `<Space />` here and there.

File: src/molecules/FileUpload/ProgressBar.native.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        return '8px';
+    }
+  },
+  progressColor({ error, color }) {
+    return error ? 'negative.960' : color;
+  },
+};
+
+const ProgressContainer = styled(View)`
+  width: 100%;
```
Comment (shubhnik) [inline src/molecules/FileUpload/ProgressBar.native.js:28]: `<Size />`

File: storybook/native/main.js [hunk 8]
```diff
@@ -37,6 +37,7 @@ configure(() => {
   require('../../src/molecules/SegmentControl/SegmentControl.stories');
   require('../../src/molecules/Snackbar/Snackbar.stories');
   require('../../src/molecules/Tabs/Tabs.stories');
+  require('../../src/molecules/FileUpload/FileUpload.stories');
```
Comment (shubhnik) [inline storybook/native/main.js:40]: alphabetical sort

File: src/molecules/FileUpload/ProgressBar.native.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  height: ${(props) => props.height};
+  border-radius: 100px;
+  max-width: 100%;
+`;
+
+const ProgressBar = ({ progress, error, size, color }) => {
+  const animation = useRef(new Animated.Value(progress));
+
+  useEffect(() => {
+    Animated.timing(animation.current, {
```
Comment (shubhnik) [inline src/molecules/FileUpload/ProgressBar.native.js:47]: ek story animation ki lgaade, maybe koi timer vgera use krke? 😛

File: src/molecules/FileUpload/FileUpload.native.js (hunk 3 above)
Comment (shridhar52war) [reply src/molecules/FileUpload/FileUpload.native.js:144]: Prop to store the file name, based on file the component is conditionally rendered

**Comment by shridhar52war** [comment]
@chaitanyadeorukhkar 
Screenshot 2021-02-08 at 1 39 13 PM

File: src/molecules/FileUpload/FileUpload.native.js [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+      setFileName(fileName);
+    } else if (!fileName) {
+      setFileName('');
+    }
+    LayoutAnimation.easeInEaseOut();
+  }, [fileName]);
+
+  const handleFilePick = async () => {
+    try {
+      const res = await DocumentPicker.pick({
```
Comment (chaitanyadeorukhkar) [inline src/molecules/FileUpload/FileUpload.native.js:58]:
```suggestion
      const file = await DocumentPicker.pick({
```

Or

```suggestion
      const document = await DocumentPicker.pick({
```

File: src/molecules/FileUpload/FileUpload.native.js [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+        type: !fileTypes
+          ? DocumentPicker.types.allFiles
+          : Object.keys(DocumentPicker.types).filter((type) => fileTypes.includes(type)),
```
Comment (chaitanyadeorukhkar) [inline src/molecules/FileUpload/FileUpload.native.js:61]: Why do we need to filter things out?

**Review by chaitanyadeorukhkar — CHANGES_REQUESTED** [review]

File: src/molecules/FileUpload/FileUpload.native.js (hunk 11 above)
Comment (shridhar52war) [reply src/molecules/FileUpload/FileUpload.native.js:61]: filter allows only specific type of files to be selected. Like user wants to select only `pdf` and `doxx` from gallery, and not `.jpeg`, so jpeg files will be disabled.

**Review by shubhnik — APPROVED** [review]

**Review by shubhnik — APPROVED** [review]

**Review by chaitanyadeorukhkar — APPROVED** [review]

**Pull request merged**